File: benchmarks/bench_neuron.go
// bench_neuron.go — AI: ReLU Neuron Activation Layer
// Compile: go build -o bench_neuron_go.exe bench_neuron.go
package main

import ("fmt"; "os")

func relu(x int64) int64 {
	if x > 0 { return x }
	return 0
}

func layer(inputs, neurons int) int64 {
	var total int64
	for n := 0; n < neurons; n++ {
		var acc int64
		for i := 0; i < inputs; i++ {
			w := (int64(n)*int64(inputs) + int64(i) + 1) % 11
			x := int64(i-inputs/2) * w
			acc += relu(x)
		}
		total += acc
	}
	return total
}

func main() {
	var result int64
	for rep := 0; rep < 2000; rep++ {
		result = layer(128, 256)
	}
	fmt.Println(result)
	os.Exit(0)
}

File: benchmarks/bench_neuron.c
// bench_neuron.c — AI: ReLU Neuron Activation Layer — self-timed
#include <stdio.h>
#include <windows.h>

static inline long long relu(long long x) { return x > 0 ? x : 0; }

long long layer(int inputs, int neurons) {
    long long total = 0;
    for (int n = 0; n < neurons; n++) {
        long long acc = 0;
        for (int i = 0; i < inputs; i++) {
            long long w = ((long long)n * inputs + i + 1) % 11;
            long long x = (i - inputs / 2) * w;
            acc += relu(x);
        }
        total += acc;
    }
    return total;
}

int main(void) {
    LARGE_INTEGER freq, t0, t1;
    QueryPerformanceFrequency(&freq);
    QueryPerformanceCounter(&t0);
    long long result = 0;
    for (int rep = 0; rep < 2000; rep++)
        result = layer(128, 256);
    QueryPerformanceCounter(&t1);
    double ms = (double)(t1.QuadPart - t0.QuadPart) * 1000.0 / freq.QuadPart;
    printf("%lld\n", result);
    fflush(stdout);
    fprintf(stderr, "COMPUTE_MS=%.3f\n", ms);
    return 0;
}
